apps/docs-website/blog/2025-01-5-inversion-of-control.mdx

```diff
@@ -0,0 +1,162 @@
+
+:::info
+
+I know, Inversion of Control is an expression that can have many different meanings in computer programming, depending on the context, but in this article I'm using it in loose terms to describe the process of giving the user more control over the library's behavior.
+
+:::
+
+## Why
+
+One of the coolest things about Sheriff is the way it encapsulates the complexity of ESLint and its ecosystem, most of the times hiding the ugly details behind the curtains. The problem is that sometimes the level of abstraction is too intrusive and can obstacolate the user's ability to customize the configuration to their needs. It's a delicate balance to strike.
+
+In this article we will explore why the Inversion of Control principle is often the key to unlock the best solution.
+
+## The `noRestrictedSyntaxOverride` option
+
+### The problem
+
+ESLint offers a rule called [`no-restricted-syntax`](https://eslint.org/docs/latest/rules/no-restricted-syntax) that allows you to disallow specific Javascript syntax features.
+
+Sheriff comes with a preconfigured `no-restricted-syntax` rule that disallows some of the most common syntax features that are considered harmful or confusing.
+
+The problem is that the rule accepts an array of features, meaning that if the user doesn't like just one of these features, they had only 2 options:
+
+- disable the rule entirely
+
+  ```ts
+  noRestrictedSyntax: 0
+  ```
+- override the rule with a new one
+
+  ```ts
+  noRestrictedSyntax: [
+    2,
+    {
+      selector: "...",
+      message: "...",
+    },
+    {
+      selector: "...",
+      message: "...",
+    },
+  ]
+  ```
+
+So in Sheriff we came up with a solution that allows the user to disable just the specific feature they don't like or append new ones.
+
+The API looked like this:
+
+```ts
+noRestrictedSyntaxOverride: {
+    adjuncts: [
+        {
+            selector: "LabeledStatement",
+            message: "...",
+        },
+        {
+            selector: "ForInStatement",
+            message: "...",
+        },
+    ],
+    allows: [
+        "LabeledStatement",
+        "ForInStatement",
+    ],
+},
+```
+
+In the `adjuncts` array, the user can add new syntax features to disallow on-top of the default ones, while in the `allows` array they can disable some of the default ones.
+
+As you can see, the offered API looked pretty complex, arbitrary, tangled and unintuitive. Both `adjuncts` and `allows` suggest at "adding" something, so it could be confusing to understand "what did what" for newcomers of the library. In short: it was not user-friendly at all.
+
+### The solution
+
+Inversion of control to the rescue.
+
+Instead of wrapping the `no-restricted-syntax` option into the Sheriff options, we let the users define and configure the `no-restricted-syntax` rule directly in their own config and offer them **composables** around it. This way the user is back in the driver seat and has full control again.
```

Comment:
```suggestion
Instead of wrapping `no-restricted-syntax` options into Sheriff’s options, we let users define and configure the `no-restricted-syntax` rule directly in their own config and then offer **composables** around it. This way, our users are back in the driver’s seat and have full control once more.
```